python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -35,16 +46,194 @@ def __init__(self, params, lr):
+        updating the optimizer as needed..
+
+        Parameters
+        ----------
+        local_state_dict : dict
+            Optimizer state; should be an object returned
+            from a call to local_state_dict()
+
+        See Also
+        --------
+        local_state_dict
+        """
+        for emb_name, emb_state in local_state_dict["emb_states"].items():
+            idx = emb_state["ids"]
+            states = (
+                list(self._state[emb_name])
+                if isinstance(self._state[emb_name], tuple)
+                else [self._state[emb_name]]
+            )
+            if len(emb_state["states"]) != len(states):
+                raise ValueError(
+                    f"loaded state dict has a different number of states"
+                    f" of embedding {emb_name}"
+                )
+            name_to_index = {
+                state.name: index for index, state in enumerate(states, 0)
+            }
+            for name, state in emb_state["states"].items():
+                if name not in name_to_index:
+                    raise ValueError(
+                        "loaded state dict contains a state {name}"
+                        "that can't be found in the optimizer states"
+                    )
+                state_idx = name_to_index[name]
+                state = state.to(states[state_idx].dtype)
+                state = state.to(th.device("cpu"))
+                states[state_idx][idx] = state
+        self._defaults.update(local_state_dict["params"])
+        self.__dict__.update(local_state_dict["params"])
+
+    def save(self, f):
+        """Save the local state_dict to disk on per rank.
+
+        Saved dict contains three parts:
+
+        * 'ids': list of global ids of the nodes/edges selected locally
+        * 'params': hyper parameters of the optimizer
+        * 'states': partial optimizer states indexed by ```ids``` of entire states
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, PathLike]
+            The basic path of the file, client id
+            will be attached as full name.
+
+        See Also
+        --------
+        load
+        """
+        num_clients = dgl.distributed.get_num_client()
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+        client_id = dgl.distributed.get_rank()
+        f = f if isinstance(f, str) else str(f, "UTF-8")
+        f = f"{f}_{client_id}"
+        th.save(self.local_state_dict(), f)
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+
+    def load(self, f):
+        """Load the local state of the optimizer from the file on per rank.
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, PathLike]
+            The basic path of the file, client id
+            will be attached as full name.
```

Comment:
Same issue.

Reply:
Use ```os.PathLike``` instead. 
```pathlib``` is only one implmentation of ```PathLike```, but the interface accept returning ```str``` or ```bytes```, so we should take care of that.